### PR TITLE
Genome metadata

### LINKED ISSUE #71: Genome metadata
### Description of feature

Improve genome metadata fetching and processing

### PR DESCRIPTION
**Pull request by BethYates**
I've created a basic subworkflow that will fetch meta data for a genome assembly and process it for inclusion in a genome note. Currently this subworkflow runs a single module 'run_wget' that takes a list of urls and fetches the related data using a wget command. To run this subworkflow I've created a couple of temporary files /test.nf and workflows/test.nf these will be deleted when the subworkflow is integrated into the main pipeline. 

To run the code I've been using the command

nextflow run test.nf -profile sanger,test,singularity
 

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
  - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/genomenote/tree/master/.github/CONTRIBUTING.md)
  - [ ] If necessary, also make a PR on the nf-core/genomenote _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: assets/genome_metadata_template.csv [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#File Source, File Type, Url, Output Type
```
Comment (priyanka-surana) [inline assets/genome_metadata_template.csv:1]: Will the space in the column names cause issues?

File: nextflow_schema.json [hunk 2]
```diff
@@ -34,6 +34,28 @@
...
+                    "description": "The bioproject accesion linked to the genome assembly, for example: PRJEB49353."
+                },
+                "biosample": {
+                    "type": "string",
+                    "description": "The biosample accesion(s) linked to the samples in the experiment, for example: SAMEA7524400."
+                },
+                "genome_metadata_file_template": {
+                    "type": "string",
+                    "format": "directory-path",
+                    "description": "Path to BUSCO lineage directory. Please use absolute paths, for example, /lustre/scratch124/tol/projects/darwin.",
```
Comment (priyanka-surana) [inline nextflow_schema.json:56]: Is this description wrong or is it the same as `lineage_db`?

File: subworkflows/local/genome_metadata.nf [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                source: row.source, 
+                type: row.type, 
+                ext: row.ext, 
+            ],
+            // url 
+            row.url
+                .replaceAll(/ASSEMBLY_ACCESSION/, params.assembly)
+                .replaceAll(/TAXONOMY_ID/, params.taxon_id)
+                .replaceAll(/BIOPROJECT_ACCESSION/, params.bioproject)
+                .replaceAll(/BIOSAMPLE_ACCESSION/, params.biosample)
```
Comment (priyanka-surana) [inline subworkflows/local/genome_metadata.nf:31]: Ideally, convert the `params` to channel in the `workflow/genomenote.nf` or in your case `workflow/test.nf`.

File: subworkflows/local/genome_metadata.nf [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    // Fetch files
+    RUN_WGET ( file_list ) 
+
+    ch_versions = ch_versions.mix(RUN_WGET.out.versions.first())
+
+    // Change this to branch code to manage passing of downloaded files to the appropriate parsing module    
+    //ch_all_files = Channel.empty()
+    //.mix( RUN_WGET.out.file_path)
+
+    emit:
```
Comment (priyanka-surana) [inline subworkflows/local/genome_metadata.nf:44]: Are you not expecting any outputs?

File: workflows/test.nf [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    // SUBWORKFLOW: Read in samplesheet/paramters and validate 
+    //
+    Channel.of ( inputs ).set { ch_input }
+    // need new module to do this
+
+    //
+    // SUBWORKFLOW: Read in samplesheet, validate and stage input files
+    //
+
+    ch_file_list = Channel.fromPath("${projectDir}/assets/genome_metadata_template.csv")
```
Comment (priyanka-surana) [inline workflows/test.nf:60]: I would say define a `params.template` and set this as the default value in nextflow.config. You can use the parameter to create the channel here.

File: workflows/test.nf [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+*/
+
+def summary_params = NfcoreSchema.paramsSummaryMap(workflow, params)
+
+// Validate input parameters
+WorkflowGenomenote.initialise(params, log)
+
+// Check mandatory parameters
+if (params.assembly && params.taxon_id && params.bioproject && params.biosample) { inputs = [ params.input, params.project, params.bioproject, params.biosample ] }
```
Comment (priyanka-surana) [inline workflows/test.nf:13]: Am I understanding this wrong or is there a typo?
You do not define `input` or `project` but you assign them to `inputs`. Where are they coming from?

**Review by priyanka-surana — COMMENTED** [review]
Some formatting and structure ideas. But mostly looks good. And I was able to run the test case 👍

File: assets/genome_metadata_template.csv (hunk 1 above)
Comment (BethYates) [reply assets/genome_metadata_template.csv:1]: It shouldn't do as the code that processes this file ignores the first line, but I'll change it to be on the safe sidee.

File: nextflow_schema.json (hunk 2 above)
Comment (BethYates) [reply nextflow_schema.json:56]: I meant to remove this, I changed the code so this is no longer user input. I've removed this section

File: subworkflows/local/genome_metadata.nf (hunk 4 above)
Comment (BethYates) [reply subworkflows/local/genome_metadata.nf:44]: I am, but as the output files will be provided by a later stage in the workflow I didn't add this yet. Is this something I should be doing now?

File: workflows/test.nf (hunk 6 above)
Comment (BethYates) [reply workflows/test.nf:13]: Typos as you thought, now fixed.

File: subworkflows/local/genome_metadata.nf (hunk 3 above)
Comment (BethYates) [reply subworkflows/local/genome_metadata.nf:31]: I tried this originally but then I wasn't sure how to get the parameters back out of the channel for use in the replaceAll. I tried using ch.toList() to do this but couldn't get it working, how would you suggest doing this?

**Comment by muffato** [comment]
Re. the "nf-core linting / nf-core" CI error, bump `python-version` in `.github/workflows/linting.yml` to 3.7 as in the [BlobToolKit pipeline](https://github.com/sanger-tol/blobtoolkit/blob/dev/.github/workflows/linting.yml#L79), and that should work

File: test.nf
Comment (priyanka-surana) [inline test.nf:1]: Do you still need this file?

**Review by priyanka-surana — APPROVED** [review]

File: subworkflows/local/genome_metadata.nf (hunk 3 above)
Comment (BethYates) [reply subworkflows/local/genome_metadata.nf:31]: After discussion it was decided to leave this code as is for now

**Pull request merged**